File: sdk-install/smoke.mjs
// The installed Node SDK, driven against the binary installed beside it.
//
// Run from a fresh project holding only what `npm install @onemessagebus/sdk`
// puts there, at the version a release would stamp: the SDK resolves the binary
// through the `onemessagebus-cli` launcher it depends on, holds it to the version
// it pins, and answers once over each transport.

import { mkdtempSync, rmSync } from "node:fs";
import { tmpdir } from "node:os";
import { join } from "node:path";

import { Client, ResidentTransport, SDK_VERSION } from "@onemessagebus/sdk";

const expected = process.argv[2];
if (SDK_VERSION !== expected) {
  throw new Error(`the installed SDK is ${SDK_VERSION}, and this revision packs ${expected}`);
}

const scratch = mkdtempSync(join(tmpdir(), "sdk-install-"));
try {
  const config = { transportDir: join(scratch, "channel") };
  const oneShot = new Client({ config });
  const kinds = await oneShot.transports();
  if (!kinds.some((kind) => kind.kind === "local")) {
    throw new Error("the installed binary lists no local transport");
  }
  await oneShot.send("surfaces", {
    kind: "finding",
    message: "installed",
    source: "proposal",
    blocking: false,
  });
  await oneShot[Symbol.asyncDispose]();

  const resident = new Client({
    config,
    transport: new ResidentTransport({ socket: join(scratch, "bus.sock") }),
  });
  const statuses = await resident.status("surfaces");
  await resident[Symbol.asyncDispose]();
  if (statuses[0]?.records !== 1) {
    throw new Error(`the resident core reads ${JSON.stringify(statuses[0])} for what was sent`);
  }
  console.log(
    `onemessagebus ${expected}: the installed Node SDK drove the installed binary one-shot and through a resident core`,
  );
} finally {
  rmSync(scratch, { recursive: true, force: true });
}

File: sdk-install/smoke.py
"""The installed Python SDK, driven against the binary installed beside it.

Run from a fresh virtualenv holding only what `pip install onemessagebus` puts
there, at the version a release would stamp: the SDK finds the binary its
`onemessagebus-cli` dependency installed, holds it to the version it pins, and
answers once over each transport.
"""

import asyncio
import sys
import tempfile
from pathlib import Path

import onemessagebus
from onemessagebus import Client, ClientConfig, ResidentTransport
from onemessagebus.models import TransportHello

HELLO = TransportHello.model_validate(
    {"protocol": "onemessagebus-transport", "version": 1, "config": {"kind": "installed"}}
)


async def main(expected: str) -> None:
    if onemessagebus.__version__ != expected:
        raise SystemExit(
            f"the installed SDK is {onemessagebus.__version__}, and this revision packs {expected}"
        )
    with tempfile.TemporaryDirectory() as scratch:
        declared = Path(scratch, "onemessagebus.yaml")
        declared.write_text(
            "version: 1\n"
            f"transport: {{kind: local, dir: {Path(scratch, 'bus')}}}\n"
            "queues:\n"
            "  hellos: {schema: onemessagebus.transport-hello@1}\n",
            encoding="utf-8",
        )
        config = ClientConfig(config=declared)
        async with Client(config) as client:
            if not any(kind.kind == "local" for kind in await client.transports()):
                raise SystemExit("the installed binary lists no local transport")
            await client.send("hellos", HELLO)
        resident = ResidentTransport(Path(scratch, "bus.sock"))
        async with Client(config, transport=resident) as client:
            statuses = await client.status("hellos")
            if statuses[0].records != 1:
                raise SystemExit(f"the resident core reads {statuses[0]} for what was sent")
    print(
        f"onemessagebus {expected}: the installed Python SDK drove the installed binary "
        "one-shot and through a resident core"
    )


asyncio.run(main(sys.argv[1]))
